sotodlib/io/bookbinder2.py

```diff
@@ -0,0 +1,931 @@
+    data : dict
+        Dict containing ancillary data timestreams read from HK files. This
+        will be populated after preprocess.
+    times : np.ndarray
+        Timestamps for anc data. This will be populated after preprocess.
+    anc_frame_data : List[G3TimestreamMap]
+        List of G3TimestreamMaps saved for each bound frame. This will be
+        populated on bind and should be used to add copies of the anc data to
+        the detector frames. 
+    """
+    def __init__(self, files, book_id, log=None):
+        self.files = files
+        self.times = None
+        self.data = None
+        self.anc_frame_data = None
+        self.out_files = []
+        self.book_id = book_id
+
+        if log is None:
+            self.log = logging.getLogger('bookbinder')
+        else:
+            self.log = log
+    
+    def preprocess(self, start=None, stop=None):
+        """
+        Preprocesses HK data and populates the `data` and `times` objects.
+        """
+        self.log.info("Preprocessing HK Data")
+        frame_iter = get_frame_iter(self.files)
+
+        data = {}
+        times = []
+        block_idx = None
+        for fr in frame_iter:
+            if fr['hkagg_type'] != 2:
+                continue
+            if 'ACU_broadcast' not in fr['block_names']:
+                continue
+            if block_idx is None:
+                block_idx = list(fr['block_names']).index('ACU_broadcast')
+
+            block = fr['blocks'][block_idx]
+            ts = np.array(block.times) / core.G3Units.s
+            if start is not None:
+                if ts[-1] < start:
+                    continue
+            if stop is not None:
+                if ts[0] >= stop:
+                    continue
+
+            times.append(ts)
+            for k, v in block.items():
+                if k not in data:
+                    data[k] = []
+                data[k].append(v)
+        
+        if not times:
+            times = np.array([], dtype=np.float64)
+            data = {}
+        else:
+            times = np.hstack(times)
+            for k, v in data.items():
+                data[k] = np.hstack(v)
+
+        self.data = data
+        self.times = times
+    
+    def bind(self, outdir, times, frame_idxs, file_idxs):
+        """
+        Binds ancillary data.
+
+        Params
+        ----------
+        outdir : path
+            Path where files should be written
+        times : np.ndarray
+            Array of timestamps to write to book
+        frame_idxs : np.ndarray
+            Array mapping sample (in times) to output frame index
+        file_idxs : np.ndarray
+            Array mapping output frame idx to output file idx
+        """
+        self.log.info("Binding ancilary data")
+
+        # Handle file writers
+        writer = None
+        cur_file_idx = None
+        out_files = []
+
+        az, el, boresight, corotation = None, None, None, None
+        if 'Corrected_Azimuth' in self.data:
```

Comment:
what are these corrected fields?

Reply:
Lauren’s explanation was:
>Within the `ACU_position` block, all fields labeled `Raw` are raw values from the encoders, and all fields labeled `Corrected` are encoder values that may have a correction factor applied to them from the ACU pointing model.

She also told me to use the corrected values.

Reply:
Although the ACU supports a rich TPOINT+ model for Raw -> Corrected, my expectation is that we will dial in, at most, only the main "encoder offsets", determined early on in commissioning.

If somehow Corrected becomes weird, we are recording the Raw & Corrected in the HK so we'll be able to figure out what happened and undo it in post.

I am actually quite open to the idea that we just use Raw fields everywhere... it sort of depends on how bad the Raw pointing is.  If az/el encoder offsets are large, but that's all we need, then there will be some pressure to set it and forget it.  But if the pointing is super weird, and encoder offsets only partly help with that, then we might as well not bother setting them (in which case Raw = Encoder).

The reasons I am against enabling the ACU's full pointing model: (a) some parameters will cause mixing of az and el, which we don't want to do while scanning (a "constant el" scan should be one where the elevation motor does not move); (b) we care enough about pointing that we'll need to measure and refine our physical pointing model very carefully over time, and apply it in a way that properly handles parallactic rotation, and if the ACU is doing non-trivial corrections we'll just have to undo those before applying our physical model to characterize the system.

Reply:
Which is the more appropriate one for use in determining telescope states/turnarounds? I’m a little concerned about the “mixing of az and el”: the Bookbinder currently tags any v_el > 10^-4 deg/sec as a slew instead of a scan. (Bookbinder has access to both Raw and Corrected so it doesn’t matter too much which is used vs which is written, but we should decide how it’ll be done.)